Python: Get most CI green

lucasmcdonald3 opened the pull request:
*Issue #, if available:*

*Description of changes:*

This gets most CI green. (Caveat: it does this by having separate transpile commands for NET/Java vs Python, but this SHOULD only be separate until NET/Java get on 4.8.)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

robin-aws committed:
Add more operations to ddb-lite, fix a small codegen bug (#520)

lucasmcdonald3 committed:
sync

robin-aws commented on .github/workflows/smithy-polymorph.yml:
```diff
@@ -31,11 +31,17 @@ jobs:
           arguments: :codegen-client:pTML :codegen-core:pTML :rust-runtime:pTML
           build-root-directory: smithy-dafny-codegen-modules/smithy-rs
+      
+      - name: Install smithy-dafny-codegen Python dependencies locally
```
Just FYI I had to do this because `smithy-rs` packages aren't published at all yet, but I assume you have to do it here because you're still patching the `smithy-python` code?

robin-aws replied:
Also it might be worth a reusable `.github/actions` for this since it's not very DRY.

lucasmcdonald3 replied:
Yes and yes (both because Smithy-Python isn't on Maven *and* because I have patches.)

lucasmcdonald3 replied:
I'm looking at this and don't see a *ton* of value in abstracting this to a resuable GHA?
I might hold off on this for now unless you think it's blocking.

lucasmcdonald3 replied:
Made a GHA for "Install Smithy-Dafny codegen dependencies"

robin-aws replied:
FYI I snuck these targets in for human usage too: https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/Makefile#L13 :)

Probably good to add the python dependency there too, perhaps in this PR so we don't forget?

robin-aws commented on codegen/settings.gradle.kts:
```diff
@@ -5,9 +5,6 @@ rootProject.name = "smithy-dafny"
 include(":smithy-dafny-codegen")
 include(":smithy-dafny-codegen-cli")
 include(":smithy-dafny-codegen-test")
-// TODO: Once Smithy-Python is published to Maven, and we do not rely on a fork, use that
```
This should probably be tracked in an issue for both Python and Rust

lucasmcdonald3 replied:
You mean an issue in Smithy-Python/rs that says "Publish codegen to Maven"? I agree

lucasmcdonald3 replied:
Python: https://github.com/smithy-lang/smithy-python/issues/296
Rust already had one: https://github.com/smithy-lang/smithy-rs/issues/421

robin-aws commented on codegen/smithy-dafny-codegen-test/src/test/java/software/amazon/polymorph/smithypython/PythonTestModels.java:
```diff
@@ -0,0 +1,47 @@
+        DISABLED_TESTS.add("SimpleTypes/SimpleFloat");
+        DISABLED_TESTS.add("SimpleTypes/SimpleShort");
+        DISABLED_TESTS.add("SimpleTypes/SimpleTimestamp");
+        DISABLED_TESTS.add("aws-sdks/ddb-lite");
```
FYI the -lite models should be very easy to turn on as well, since the projection is handled before hitting any language-specific logic.

lucasmcdonald3 replied:
Took a look, I'll leave these unimplemented for now.
Even though the models are "lite", they use the gradle build process, which seems to need extra work to get working.
I think Python and Go will need extra work to figure out how to pass in extra args for the "dependency-module-names" and "library-name" flags.

robin-aws replied:
Fair enough, there's no real motivation to get those working now anyway - they are mainly useful during the development of new languages and you're already past that stage.

robin-aws commented on SmithyDafnyMakefile.mk:
```diff
@@ -156,9 +156,16 @@ clean-dafny-report:
+_transpile_implementation_all: TRANSPILE_DEPENDENCIES=$(patsubst %, -library:$(PROJECT_ROOT)/%, $(PROJECT_INDEX))
 _transpile_implementation_all: transpile_implementation
 
+# TODO-new-cli: As part of supporting Dafny 4.8+, Java and NET SHOULD migrate to the new Dafny CLI,
```
This feels slightly off and could be more precise - is it because the legacy CLI will eventually be fully deprecated? It seems more like other languages need features that are only in the new CLI.

lucasmcdonald3 replied:
Yes, but the more immediate desire is that Java/NET/Python/Go have the same build command.
This reduces risk of accidental "forks" in one build process or the other.
(I know we will need some Makefile checking that says "only add `--compile-suffix:1` for Java/NET", but that is a small "fork".)

This is meant to assert that "we have a 4.8 upgrade coming up, these targets SHOULD be merged as part of that upgrade"

robin-aws commented on SmithyDafnyMakefile.mk:
The `_new_cli` duplication seems reasonable to me. Only question is whether this PR also ensures they are as well tested. It seems like it is since Python will only use them?

lucasmcdonald3 replied:
Yes, right now only Python.
My goal is to get this into main soon so Rishav and Shubham can onboard to a "shared" solution.

robin-aws commented on TestModels/Aggregate/src/Index.dfy:
```diff
@@ -2,7 +2,7 @@
 include "SimpleAggregateImpl.dfy"
 
-module SimpleAggregate refines AbstractSimpleAggregateService {
+module {:extern "simple.aggregate.internaldafny"} SimpleAggregate refines AbstractSimpleAggregateService {
```
Is this a breaking change for existing projects?

lucasmcdonald3 replied:
Nope! This is the current state of main-1.x.
I'm putting the extern names back in the python-reviewed branch so that when I merge this branch in, there is no diff between its Dafny source and main's.

robin-aws replied:
Ah I actually missed that this is not targeting `main-1.x` itself.

robin-aws commented on TestModels/CodegenPatches/runtimes/python/pyproject.toml:
Consider implementing `--generate project-files` for Python in the future, so many libraries don't have to provide this file.

lucasmcdonald3 replied:
Good to know, thanks. I didn't realize we had this

lucasmcdonald3 committed:
gha

lucasmcdonald3 committed:
gha

ajewellamz committed:
Benerate SimpleErrors (#440)

robin-aws committed:
chore(tests): Include @dafnyUTF8Bytes in SimpleString test model (#539)

Corrected GetStringUTF8 to use this trait
Corrected the description of the existing test that incorrectly said the string was in UTF-8 (it's a Dafny string so it's implicitly UTF-16 instead)
Added identically structured test of GetStringUTF8
Updated benerated Rust code to make sure this case is handled too.

lucasmcdonald3 committed:
Merge branch 'main-1.x' into ci

lucasmcdonald3 committed:
consolidate
